src/libraries/System.Net.NameResolution/src/System/Net/NameResolutionPal.Windows.cs

```diff
@@ -374,6 +388,59 @@ public GetAddrInfoExState(string hostName, bool justAddresses)
 
             public Task Task => JustAddresses ? (Task)IPAddressArrayBuilder.Task : IPHostEntryBuilder.Task;
 
+            public void RegisterForCancellation(CancellationToken cancellationToken)
+            {
+                if (!cancellationToken.CanBeCanceled) return;
+
+                lock (this)
+                {
+                    if (_cancellationContext == null)
+                    {
+                        // The operation completed before registration could be done.
+                        return;
+                    }
+
+                    _cancellationRegistration = cancellationToken.UnsafeRegister(o =>
+                    {
+                        var @this = (GetAddrInfoExState)o!;
+                        int cancelResult = 0;
+
+                        lock (@this)
+                        {
+                            GetAddrInfoExContext* context = @this._cancellationContext;
+
+                            if (context != null)
+                            {
+                                // An outstanding operation will be completed with WSA_E_CANCELLED, and GetAddrInfoExCancel will return NO_ERROR.
+                                // If this thread has lost the race between cancellation and completion, this will be a NOP
+                                // with GetAddrInfoExCancel returning WSA_INVALID_HANDLE.
+                                cancelResult = Interop.Winsock.GetAddrInfoExCancel(&context->CancelHandle);
+                            }
+                        }
+
+                        if (cancelResult != 0 && NetEventSource.IsEnabled)
+                        {
+                            string error =
+                                cancelResult == Interop.Winsock.WSA_INVALID_HANDLE ? "GetAddrInfoExCancel lost a race with GetAddrInfo; unable to cancel already completed operation." :
```

Comment:
We generally haven't logged such races in other places, and they're possible all over the place.  Is it really important here?  If it were me, I'd instead do:
```C#
if (cancelResult != 0 && NetEventSource.IsEnabled && cancelResult != Interop.Winsock.WSA_INVALID_HANDLE)
{
    NetEventSource.Info(@this, $"GetAddrInfoExCancel returned an error: {cancelResult}");
}
```

Reply:
I'm fine leaving it out.